stdlib/http/src/main/ballerina/http/http_secure_client.bal

```diff
@@ -294,178 +345,603 @@ public function createHttpSecureClient(string url, ClientEndpointConfig config)
+# + return - Prepared HTTP request object
+function prepareRequest(RequestConfig config) returns Request|error {
+    Request req = new;
+    string textPayload = config.payload;
+    string scopeString = EMPTY_STRING;
+    string[]? scopes = config.scopes;
+    if (scopes is string[]) {
+        foreach var requestScope in scopes {
+            scopeString = scopeString + WHITE_SPACE + requestScope;
+        }
+    }
+    if (scopeString != EMPTY_STRING) {
+        textPayload = textPayload + "&scope=" + scopeString.trim();
+    }
+
+    var clientId = config["clientId"];
+    var clientSecret = config["clientSecret"];
+    if (config.credentialBearer == AUTH_HEADER_BEARER) {
+        if (clientId is string && clientSecret is string) {
+            string clientIdSecret = clientId + ":" + clientSecret;
+            req.addHeader(AUTH_HEADER, AUTH_SCHEME_BASIC + WHITE_SPACE +
+                    encoding:encodeBase64(clientIdSecret.toByteArray("UTF-8")));
+        } else {
+            string errMsg = "Client ID or client secret is not provided for client authentication.";
+            log:printError(errMsg);
+            error e = error(HTTP_ERROR_CODE, { message: errMsg });
+            return e;
+        }
+    } else if (config.credentialBearer == POST_BODY_BEARER) {
+        if (clientId is string && clientSecret is string) {
+            textPayload = textPayload + "&client_id=" + clientId + "&client_secret=" + clientSecret;
+        } else {
+            string errMsg = "Client ID or client secret is not provided for client authentication.";
+            log:printError(errMsg);
+            error e = error(HTTP_ERROR_CODE, { message: errMsg });
+            return e;
+        }
+    }
+    req.setTextPayload(untaint textPayload, contentType = mime:APPLICATION_FORM_URLENCODED);
+    return req;
+}
+
+# Extract access token from the json payload of given HTTP response and update the token cache.
+#
+# + response - HTTP response object
+# + tokenCache - Cached token configurations
+# + clockSkew - Clock skew in seconds
+# + return - Extracted access token or `error` if error occured during HTTP client invocation
+function extractAccessTokenFromResponse(Response response, CachedTokenConfig tokenCache,
+                                        int clockSkew) returns string|error {
+    if (response.statusCode == OK_200) {
+        var payload = response.getJsonPayload();
+        if (payload is json) {
+            log:printDebug(function () returns string {
+                    return "Received an valid response. Extracting access token from the payload: " + <string>payload;
+                });
+            check updateTokenCache(payload, tokenCache, clockSkew);
+            return payload.access_token.toString();
+        } else {
+            string errMsg = "Failed to retrieve access token since the response payload is not a JSON.";
+            log:printError(errMsg);
+            error e = error(HTTP_ERROR_CODE, { message: errMsg });
```

Comment:
Would it make sense in such scenarios to include the original error or at least the reason string from the original error in the new error's detail map?